most-used-words.el

```diff
@@ -50,97 +50,170 @@
-    (if show-percentages-p
-	(list (cl-subseq sorted-counts 0 n) total-count)
-      (mapcar #'cl-first (cl-subseq sorted-counts 0 n)))))
+Optional argument SHOW-PERCENTAGES-P displays word counts and percentages."
+  (with-current-buffer most-used-words--buffer
+    (let ((counts (make-hash-table :test #'equal))
+          (total-count 0) sorted-counts
+          start end)
+      (save-excursion
+        (goto-char (point-min))
+        (skip-syntax-forward "^w")
+        (setf start (point))
+        (cl-loop until (eobp)
+                 do
+                 (skip-syntax-forward "w")
+                 (setf end (point))
+                 (cl-incf (gethash (buffer-substring start end) counts 0))
+                 (skip-syntax-forward "^w")
+                 (setf start (point))
+                 (cl-incf total-count)))
+      (cl-loop for word being the hash-keys of counts
+               using (hash-values count)
+               do
+               (push (list word count) sorted-counts)
+               finally (setf sorted-counts (cl-sort sorted-counts #'>
+                                                    :key #'cl-second)))
+      (if show-percentages-p
+          (list (cl-subseq sorted-counts 0 n) total-count)
+        (mapcar #'cl-first (cl-subseq sorted-counts 0 n))))))
 
 (defun most-used-words-buffer-aux (&optional n show-counts-p)
   "Show the N (default 3) most used words in the current buffer.
 
-Optional argument SHOW-COUNTS-P also shows the counts and
-percentages."
-  (let ((most-used-words-buffer (get-buffer-create "*Most used words*")))
-    (if show-counts-p
-	(let ((most-used (most-used-words-buffer-1 n t)))
-	  (most-used-words-with-view-buffer most-used-words-buffer
-            (let ((word-counts (cl-first most-used))
-		  (total-count (float (cl-second most-used))))
-	      (cl-loop for (word count) in word-counts
-		 do
-		   (insert (format "%-24s    %5d    %f%%" word count (* 100 (/ count total-count))))
-		   (newline)))))
-      (let ((most-used (most-used-words-buffer-1 n)))
-	(most-used-words-with-view-buffer most-used-words-buffer
-          (dolist (word most-used)
-	    (insert (format "%s" word))
-	    (newline)))))))
+Optional argument SHOW-COUNTS-P also shows the counts and percentages."
+  (let* ((most-used-words--buffer (current-buffer))
+         (most-used-words-buffer (get-buffer-create most-used-words--buffer-name))
+         (most-used (most-used-words-data n))
+         word count percent)
+    (cl-case most-used-words-display-type
+      ('buffer
+       (most-used-words-with-view-buffer
+        most-used-words-buffer
+        (dolist (word-data most-used)
+          (setq word (plist-get word-data :word)
+                count (plist-get word-data :count)
+                percent (plist-get word-data :percent))
+          (if show-counts-p
+              (insert (format "%-24s    %5d    %0.2f%%" word count percent))
+            (insert (format "%s" word)))
+          (newline))))
```

Comment:
I have simplify the insertion of the buffer. I assume this reduced much complexity to this function. Any idea? 😖

Reply:
Seems fine.